File: front-end/src/app/domain/models/game/action/impl/property/LocationPurchaseAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";
import {OwnableLocation} from "../../../board/location/OwnableLocation";
import {AssetPurchaseAction} from "../money/AssetPurchaseAction";
import {LocationUtil} from "../../../board/location/LocationUtil";

export class LocationPurchaseAction extends AssetPurchaseAction {
  private _asset: OwnableLocation;

  public constructor(
    id: number,
    buyer: Player,
    performedAt: Date,
    asset: OwnableLocation,
    previousOwner: Player,
    purchasePrice: number
  ) {
    super(LocationPurchaseAction.name, id, buyer, previousOwner, performedAt, purchasePrice);
    this._asset = asset;
  }

  public get asset(): OwnableLocation {
    return this._asset;
  }

  public static trueCopy(action: LocationPurchaseAction): GameAction | null {
    return new LocationPurchaseAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      <OwnableLocation>LocationUtil.trueCopy(action.asset),
      action.recipient == null ? null : Player.trueCopy(action.recipient),
      action.amount,
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "asset": this.asset.toJson(),
      }
    );
  }

  public override toString(): string {
    return `${this.actor?.user.username} bought ${this.asset.toString()} from ${this.recipient == null ? "the bank" : this.recipient.user.username}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/property/LocationPurchaseAction.java
package nl.hva.backend.domain.models.game.action.impl.property;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl.AssetPurchaseAction;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

/**
 * Representation, of the action, of purchasing a location within the game.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class LocationPurchaseAction extends AssetPurchaseAction {

    @ManyToOne
    private OwnableLocation asset;

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param buyer         The player, or bank, who this property belongs to from now on.
     * @param previousOwner The player, or bank, who this property belonged to.
     * @param purchasePrice The cost of the purchase.
     * @param asset         The OwnableLocation being purchased.
     */
    public LocationPurchaseAction(
            Player buyer,
            Player previousOwner,
            double purchasePrice,
            OwnableLocation asset
    ) {
        super(buyer, previousOwner, purchasePrice);
        this.asset = asset;
    }

}